language/move-core/types/src/language_storage.rs

```diff
@@ -79,6 +103,153 @@ impl<'a> From<&'a TypeTag> for SafeTypeTag<'a> {
+            DeTypeTag::Bool => {
+                variant.unit_variant()?;
+                Ok(TypeTag::Bool)
+            }
+            DeTypeTag::U8 => {
+                variant.unit_variant()?;
+                Ok(TypeTag::U8)
+            }
+            DeTypeTag::U64 => {
+                variant.unit_variant()?;
+                Ok(TypeTag::U64)
+            }
+            DeTypeTag::U128 => {
+                variant.unit_variant()?;
+                Ok(TypeTag::U128)
+            }
+            DeTypeTag::Address => {
+                variant.unit_variant()?;
+                Ok(TypeTag::Address)
+            }
+            DeTypeTag::Signer => {
+                variant.unit_variant()?;
+                Ok(TypeTag::Signer)
+            }
+            DeTypeTag::Vector => Ok(TypeTag::Vector(Box::new(
+                variant.newtype_variant_seed(NestedTypeTag(self.0))?,
+            ))),
+            DeTypeTag::Struct => Ok(TypeTag::Struct(
+                variant.newtype_variant_seed(NestedStructTag(self.0))?,
+            )),
+        }
+    }
+}
+
+// u8 -> depth or nesting level
+struct NestedTypeTag(u8);
+
+impl<'de> DeserializeSeed<'de> for NestedTypeTag {
+    type Value = TypeTag;
+
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        if self.0 >= MAX_DE_TYPE_TAG_NESTING {
+            return Err(de::Error::custom(format!(
+                "Exceeded TypeTag nesting limit during deserialization: {}",
+                self.0
+            )));
+        }
+
+        deserializer.deserialize_enum("TypeTag", TYPE_TAG_VARIANTS, TypeTagVisitor(self.0 + 1))
+    }
+}
+
+// u8 -> depth or nesting level
+struct VecTypeTagVisitor(u8);
+
+impl<'de> Visitor<'de> for VecTypeTagVisitor {
+    type Value = Vec<TypeTag>;
+
+    fn expecting(&self, formatter: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        formatter.write_str("`TypeTag`")
+    }
+
+    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+    where
+        A: SeqAccess<'de>,
+    {
+        let mut data = vec![];
+        while let Some(elem) = seq.next_element_seed(NestedTypeTag(self.0))? {
+            data.push(elem);
+        }
+        Ok(data)
+    }
+}
+
+// u8 -> depth or nesting level
+struct NestedVecTypeTag(u8);
+
+impl<'de> DeserializeSeed<'de> for NestedVecTypeTag {
+    type Value = Vec<TypeTag>;
+
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        if self.0 >= MAX_DE_TYPE_TAG_NESTING {
+            return Err(de::Error::custom(format!(
+                "Exceeded TypeTag nesting limit during deserialization: {}",
+                self.0
+            )));
+        }
+
+        deserializer.deserialize_seq(VecTypeTagVisitor(self.0))
+    }
+}
+
+// u8 -> depth or nesting level
+struct NestedStructTag(u8);
+
+impl<'de> DeserializeSeed<'de> for NestedStructTag {
```

Comment:
Move this closer to the struct stuff?

Reply:
I suspect one could bikeshed a lot about the organization of this file. I might argue putting the public types at the top is more readable. I could also see an argument about relational placement.